proxy/envoy/config.go

```diff
@@ -607,6 +610,76 @@ func buildInboundListeners(mesh *proxyconfig.ProxyMeshConfig, sidecar proxy.Node
 	return listeners, clusters
 }
 
+func appendPortToDomains(domains []string, port int) []string {
+	domainsWithPorts := make([]string, len(domains), 2*len(domains))
+	copy(domainsWithPorts, domains)
+
+	for _, domain := range domains {
+		domainsWithPorts = append(domainsWithPorts, domain+":"+strconv.Itoa(port))
+	}
+
+	return domainsWithPorts
+}
+
+func buildEgressFromSidecarVirtualHostOnPort(rule *proxyconfig.EgressRule, mesh *proxyconfig.ProxyMeshConfig,
+	port *model.Port) *VirtualHost {
+	var externalTrafficCluster *Cluster
+
+	protocolToHandle := port.Protocol
+	if protocolToHandle == model.ProtocolGRPC {
+		protocolToHandle = model.ProtocolHTTP2
+	}
+	protocolSuffix := "-" + strings.ToLower(string(protocolToHandle))
+
+	if rule.UseEgressProxy {
+		externalTrafficCluster = buildOutboundCluster("istio-egress.default-"+protocolSuffix, port, nil)
+		externalTrafficCluster.ServiceName = ""
+		externalTrafficCluster.Type = ClusterTypeStrictDNS
+		externalTrafficCluster.Hosts = []Host{{URL: fmt.Sprintf("tcp://%s", mesh.EgressProxyAddress)}}
+	} else {
+		externalTrafficCluster = buildOriginalDSTCluster("orig-dst-cluster"+protocolSuffix, mesh.ConnectTimeout)
+
+		if protocolToHandle == model.ProtocolHTTPS {
+			externalTrafficCluster.SSLContext = &SSLContextExternal{}
+		}
+
+		if protocolToHandle == model.ProtocolHTTP2 {
+			externalTrafficCluster.Features = ClusterFeatureHTTP2
+		}
+	}
+
+	externalTrafficRoute := buildDefaultRoute(externalTrafficCluster)
+
+	return &VirtualHost{
+		Name:    rule.Domains[0], // use the first domain as the name
```

Comment:
name must be unique among all virtual hosts on http connection manager. Can you use the rule name/namespace plus some egress prefix to prevent accidental duplicates.

Reply:
Please note that the domains must be unique among all the virtual hosts. 
https://lyft.github.io/envoy/docs/configuration/http_conn_man/route_config/vhost.html#config-http-conn-man-route-table-vhost 
> A domain must be unique across all virtual hosts or the config will fail to load.

That means domains have to be unique anyway, and from that follows that the name of the virtual hosts will be unique as well (being the first domain of the rule).

Reply:
Ok, that is OK for now. I would like to add a TODO for later for you to make sure no two duplicate domains appear in the config. Our config backend cannot enforce a guarantee across a category of resources, so this must be done during config generation. I'm interested to get this merged soon, so I can apply the config refactor WIP to egress rules.